core/frontend/src/components/autopilot/FirmwareManager.vue

```diff
@@ -0,0 +1,269 @@
+          class="mr-2"
+          @click="toggle"
+        >
+          <v-icon class="mr-2">
+            mdi-cloud
+          </v-icon>
+          <div>Choose firmware from repository</div>
+        </v-btn>
+      </v-item>
+      <v-item
+        v-slot="{ active, toggle }"
+        :value="UploadType.File"
+      >
+        <v-btn
+          :color="active ? 'primary' : ''"
+          @click="toggle"
+        >
+          <v-icon class="mr-2">
+            mdi-paperclip
+          </v-icon>
+          <div>Upload custom firmware file</div>
+        </v-btn>
+      </v-item>
+    </v-item-group>
+
+    <v-form
+      v-if="upload_type === UploadType.Cloud"
+      ref="form"
+      lazy-validation
+    >
+      <v-select
+        v-model="chosen_vehicle"
+        :items="vehicle_types"
+        label="Vehicle"
+        required
+        @change="updateAvailableFirmwares"
+      />
+      <v-select
+        v-model="chosen_firmware_url"
+        :disabled="disable_firmware_selection"
+        :items="showable_firmwares"
+        label="Firmware"
+        required
+        @change="setCloudFirmwareAvailable"
+      />
+    </v-form>
+
+    <v-file-input
+      v-if="upload_type === UploadType.File"
+      label="Firmware file"
+      @change="setFileFirmwareAvailable"
+    />
+
+    <v-btn
+      :disabled="!allow_installing"
+      class="mt-6"
+      @click="installFirmware"
+    >
+      <v-icon>mdi-paperclip</v-icon>
+      <div>Install firmware</div>
+    </v-btn>
+
+    <v-dialog
+      v-model="installing"
+      hide-overlay
+      persistent
+      width="300"
+    >
+      <v-card
+        color="primary"
+        dark
+      >
+        <v-card-text>
+          Installing firmware. Please wait.
+          <v-progress-linear
+            indeterminate
+            color="white"
+            class="mb-0"
+          />
+        </v-card-text>
+      </v-card>
+    </v-dialog>
+
+    <v-alert
+      v-model="install_completed"
+      dismissible
+      :color="install_alert_color"
+      border="left"
+      elevation="2"
+      colored-border
+      :icon="install_alert_icon"
+      class="ma-6"
+    >
+      <p>{{ install_result_message }}</p>
+    </v-alert>
+  </v-card>
+</template>
+
+<script lang="ts">
+import axios, { AxiosRequestConfig } from 'axios'
+import Vue from 'vue'
+import { getModule } from 'vuex-module-decorators'
+
+import AutopilotStore from '@/store/autopilot'
+import NotificationStore from '@/store/notifications'
+import { Firmware, Vehicle } from '@/types/autopilot'
+import { autopilot_service } from '@/types/frontend_services'
+import { LiveNotification, NotificationLevel } from '@/types/notifications'
+
+const notification_store: NotificationStore = getModule(NotificationStore)
+const autopilot_store: AutopilotStore = getModule(AutopilotStore)
+
+enum UploadType {
+  Cloud,
+  File
+}
+
+export default Vue.extend({
+  name: 'FirmwareManager',
+  data() {
+    return {
+      UploadType,
```

Comment:
I'm not sure if I follow, what is this UploadType here ?

Reply:
This is injecting UploadType enum into the Vue instance, making it usable inside template.